lib/spack/spack/installer.py

```diff
@@ -546,14 +546,14 @@ def package_id(pkg):
     # TODO: test_force_uninstall_and_reinstall_by_hash` test.
```

Comment:
@tgamblin @becker33 These comments above are I think why @haampie thought this patch might not be accepted.  Unfortunately, 'test_force_uninstall_and_reinstall_by_hash' was the last test I attempted to fix, and not until I couldn't fix it, did I really notice this comment.

Reply:
I took a look at this and I think I see what the issue is.  `test_force_uninstall_and_reinstall_by_hash` starts with this:

```
b4ervof mpileaks@2.3
lbzbz32     callpath@1.0    <-- to be uninstalled
xvb3xy5         dyninst@8.2
wvuwyho             libdwarf@20130729
hve5eeh                 libelf@0.8.13
4v3nsfv         mpich@3.0.4
```

It force-uninstalls the callpath library there, but its spec (`lbzbz32`) is still in the DB because it's needed by `mpileaks`.  We then try to reinstall `/lbzbz32` (the exact same spec).

It looks like #15197 made it so that we skip putting `dyninst` on the queue initially, which is the right thing to do, since it's already installed, but the priority in the queue is still calculated directly from the spec's dependency count in `BuildTask` when `callpath` is enqueued.  So it looks like it still has that extra `dyninst` dependency and fails.

I think the logic from #15197 needs to be updated so that BuildTask is told what the initial dep count should be.  It seems to be assuming that everything is initially uninstalled when it should probably take the dep count as an argument, as only the algorithm that sets up the queue knows which deps should be skipped.

I don't fully understand why we fail like this:

```
spack.installer.InstallError: Cannot proceed with callpath-1.0-lbzbz32w5aiiesugnacmcht2frcksb4v: 1 uninstalled dependency: dyninst-8.2-xvb3xy52riui7jyceawrf3v3p4k5m4qk
```

I feel like if my explanation above is correct, it would say there are *two* uninstalled dependencies, not one, so I'm suspicious that I am missing something.

@tldahlgren: can you look at this one?  What am I missing?

Reply:
Following up on a refresher investigation and offline discussions, this is the "dependents" problem alluded to in the `TODO` comments of `installer.py`.  

Specifically, the re-install in the test will process a build task for `dyninst-8.2-xvb3xy52riui7jyceawrf3v3p4k5m4qk`, flag it as installed and attempt to remove it from its dependents' uninstalled dependencies.  Unfortunately, the only dependent of `dyninst` is `callpath-1.0-nwc2p47aq2tmt4w2ab5pqnrwqqztjo76` -- note the hash difference -- which does not have an associated build task to update and requeue.  Hence, the `callpath` identified in the error doesn't know that the package has been installed.